### PR TITLE
Test Harness

### LINKED ISSUE #61: Test Harness Class Skeleton
Test Harness Main Class Implementation

### LINKED ISSUE #74: Test Harness
Test harness implementation

### LINKED ISSUE #103: TestHarness_Functions_Patchup
Adding the review version of the Test Harness functions excluding the result logging

### LINKED ISSUE #107: Test Harness - Add Result Logging

### PR DESCRIPTION
**Pull request by Rahul-R5**
Contains complete implementation of test harness
Closes #107 Result logging
Closes #103 Related issue
Closes #61 Skeleton code
Closes #74 Overall

### CODE REVIEW

File: src/infrastructure/validation/testing/TestHarness.java [hunk 1]
```diff
@@ -0,0 +1,269 @@
...
+
+      File directoryPath = new File(path);
+      //get all the test cases in tests directory of respective category 
+      String tests[] = directoryPath.list();
+      totalNumberOfTests = tests.length;
+      
+      //create object of each test case class to run and get the result of each test
+      for(int i=0; i<tests.length; i++){
+        //Get full qualified class name from absolute path of testcase
+        String absPath = tests[i].getAbsolutePath();
```
Comment (convict-git) [inline src/infrastructure/validation/testing/TestHarness.java]: `getAbsolutePath()` is a method for `File` Class. You are applying it on a `String`. 
If you were programming on a Java supported IDE, it must have shown an error.

File: src/infrastructure/validation/testing/TestHarness.java [hunk 2]
```diff
@@ -0,0 +1,269 @@
...
+        if(category.equals(test.getCategory())){
+          boolean result = test.run();
+
+          if(result == false){
+            failedNumberOfTests++;
+            if(failedNumberOfTests == 1){
+              logger.log(ModuleID.TEST, LogLevel.INFO, "Failed Tests...");
+            }
+           	//log the result of failed test cases
+            logger.log(ModuleID.TEST, LogLevel.INFO, Integer.toString(failedNumberOfTests)+". "+testName+": "+test.getError());
```
Comment (convict-git) [inline src/infrastructure/validation/testing/TestHarness.java]: `testName` isn't defined at all. Neither it's a class variable nor a local variable.

File: src/infrastructure/validation/testing/TestHarness.java [hunk 3]
```diff
@@ -0,0 +1,269 @@
...
+          }
+          else{
+            successfulNumberOfTests++;
+          }   
+        }
+      } 
+		}
+    catch (Exception e){
+      logger.log(ModuleID.TEST, LogLevel.WARNING, "Exception in module: "+strModule+" >> "+e);
```
Comment (convict-git) [inline src/infrastructure/validation/testing/TestHarness.java]: Again, `strModule` isn't defined at all.

File: src/infrastructure/validation/testing/TestHarness.java [hunk 4]
```diff
@@ -0,0 +1,269 @@
...
+        if(priority == (test.getPriority())){
+           boolean result = test.run(); 
+           if(result == false){
+            failedNumberOfTests++;
+            if(failedNumberOfTests == 1){
+              logger.log(ModuleID.TEST, LogLevel.INFO, "Failed Tests...");
+            }
+            //log the result of failed test cases
+            logger.log(ModuleID.TEST, LogLevel.INFO, Integer.toString(failedNumberOfTests)+". "+testName+": "+test.getError());
+           }
```
Comment (convict-git) [inline src/infrastructure/validation/testing/TestHarness.java]: `testName` not defined :shrug:

File: src/infrastructure/validation/testing/TestHarness.java [hunk 5]
```diff
@@ -0,0 +1,269 @@
...
+      Object test = testClass.getDeclaredConstructor().newInstance();
+      boolean result = test.run();
+
+      if(result == false){
+        failedNumberOfTests++;
+        if(failedNumberOfTests == 1){
+          logger.log(ModuleID.TEST, LogLevel.INFO, "\nFailed Test...");
+        }
+        //log the result of failed test cases
+        logger.log(ModuleID.TEST, LogLevel.INFO, Integer.toString(failedNumberOfTests)+". "+testName+" :"+test.getError());
```
Comment (convict-git) [inline src/infrastructure/validation/testing/TestHarness.java]: `testName` not defined

File: src/infrastructure/validation/testing/TestHarness.java [hunk 6]
```diff
@@ -0,0 +1,269 @@
...
+      }
+    }
+    catch (Exception e){
+      logger.log(ModuleID.TEST, LogLevel.WARNING, "Exception >> "+e);      	
+    }
+      
+    //result logging 
+    logger.log(ModuleID.TEST, LogLevel.INFO, "\nOverall Result:");
+    logger.log(ModuleID.TEST, LogLevel.INFO, "Total Number of Tests: "+Integer.toString(totalNumberOfTests));
+    logger.log(ModuleID.TEST, LogLevel.INFO, "Total Number of Successful Tests: "+Integer.toString(successfulNumberOfTests));
```
Comment (convict-git) [inline src/infrastructure/validation/testing/TestHarness.java]:
```suggestion
    logger.log(
       ModuleID.TEST, 
       LogLevel.INFO, 
       "Total Number of Successful Tests: "
       + Integer.toString(successfulNumberOfTests)
    );
```

All the logs can be changed to this format.

**Review by convict-git — CHANGES_REQUESTED** [review]
@Rahul-R5 Please get the things fixed.
I'm not sure why PR coming from the Infrastructure team has compilation issues.
Is the team not using Eclipse as suggested or any other Java IDE?

And are the codes not pulled locally for testing and reviewing at L2? Is the code just briefly reviewed from Github's UI and sent for merge at L1? 

I see some wrong methods and a lot of undeclared variables. :disappointed: PR coming at L2 is supposed to be deployable.

**Comment by Rahul-R5** [comment]
@convict-git Yes, I am aware it will have errors. That's why I informed you that it is not tested yet and I didn't put up a "Ready for review" comment on this PR. Why did you assume that this was the final version and needed to be reviewed? I merely created this PR for the progress to be tracked, did not request a review yet.

**Comment by convict-git** [comment]
> Why did you assume that this was the final version and needed to be reviewed?

Multiple reasons but the main one "There was no [WIP] tag"

@Rahul-R5 Why is the PR closed?

**Comment by Rahul-R5** [comment]
> @Rahul-R5 Why is the PR closed?

Branch was resetted and it auto closed. I'll make a new one.

**Pull request closed without merging**